tests/system_tests_expandvars.py

```diff
@@ -0,0 +1,91 @@
+expandvars_script_folder = os.path.join(path_parent, 'scripts')
+script_file = "expandvars.py"
+env = dict(os.environ)
+env.update(env_config)
+
+
+class ExpandvarsTest(TestCase):
+    def test_expandvars_input_file_output_file(self):
+        # Provide an input file and an output file to the expandvars.py script
+        # and make sure the output file is populated with the expanded vars.
+        nginx_conf_dir = os.path.join(current_dir, 'nginx-configs')
+        input_file = os.path.join(nginx_conf_dir, 'expandvars.conf')
+        output_file = os.path.join(self.directory, 'expandvars.conf')
+        args = [sys.executable, os.path.join(expandvars_script_folder, script_file), input_file, output_file]
+        subprocess.run(args, stderr=STDOUT, check=True, env=env)
+        port_expanded = False
+        name_expanded = False
+        with open(output_file) as ofile:
+            for line in enumerate(ofile):
+                test_line = line[1].strip()
+                if "PORT=12345" == test_line:
+                    port_expanded = True
+                if "NAME=YoYoMa" == test_line:
+                    name_expanded = True
+        self.assertTrue(port_expanded and name_expanded)
+
+    def test_expandvars_input_file_without_output_file(self):
+        # Provide an input file and NO output file to the expandvars.py script
+        # and make sure the input file is overwritten/populated with the expanded vars.
+        nginx_conf_dir = os.path.join(current_dir, 'nginx-configs')
+        orig_file = os.path.join(nginx_conf_dir, 'expandvars.conf')
+
+        # Make a copy of tests/nginx-configs/expandvars.conf, we don't want to modify that file.
+        input_file = os.path.join(self.directory, 'expandvars.in.conf')
+        shutil.copyfile(orig_file, input_file)
+        args = [sys.executable, os.path.join(expandvars_script_folder, script_file), input_file]
+        subprocess.run(args, stderr=STDOUT, check=True, env=env)
+        port_expanded = False
+        name_expanded = False
+        with open(input_file) as ofile:
+            for line in enumerate(ofile):
+                test_line = line[1].strip()
+                if "PORT=12345" == test_line:
+                    port_expanded = True
+                if "NAME=YoYoMa" == test_line:
+                    name_expanded = True
+        self.assertTrue(port_expanded and name_expanded)
+
+    def test_expandvars_no_input_file_no_output_file(self):
+        # Does not pass any arguments to the expandvars.py
+        # and makes sure that the script fails.
+        args = [sys.executable, os.path.join(expandvars_script_folder, script_file)]
+        script_failed = False
+        try:
+            subprocess.run(args, stderr=STDOUT, check=True, env=env)
+        except CalledProcessError as e:
+            if "returned non-zero exit status 2" in str(e):
+                script_failed = True
+        self.assertTrue(script_failed)
```

Comment:
This could've been done with https://docs.python.org/3/library/unittest.html#unittest.TestCase.assertRaises

```
with self.assertRaises(CalledProcessError) as cm:
    subprocess.run(args, stderr=STDOUT, check=True, env=env)
self.assertIn("returned non-zero exit status 2", cm.exception)
```

or, maybe even better, `self.assertEquals(2, cm.exception.returncode)`, https://docs.python.org/3/library/subprocess.html#subprocess.CalledProcessError

Reply:
Done !

Reply:
Asserting on the exception text is weird and imo pointless. It would've been better to only assert on exception class and errorcode. Bit it's no big deal either way.